keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1655 @@
+   // be allocated.
+   //
+   // For an allocated resource this field defines
+   // where the resource can be used by Pods.
+   //
+   // Setting this field is optional. If nil, the
+   // resource is available everywhere.
+   AvailableOnNodes *corev1.NodeSelector
+
+   // Arbitrary data returned by the driver after a successful allocation.
+   // This data is passed to the driver for all operations involving
+   // the allocated resource. This is opaque for Kubernetes.
+   // Driver documentation may explain to users how to interpret
+   // this data if needed.
+   //
+   // The attributes must be sufficient to deallocate the resource
+   // because the ResourceClass might not be available anymore
+   // when deallocation starts.
+   Attributes map[string]string
+
+    // UsersLimit determines how many entities are allowed to use this resource
+    // at the same time. The default is 1. -1 enables the usage by an unlimited number
+   // of users. Individual containers in a pod are not counted as users, only the Pod
+   // is.
+   UserLimit int
+
+   // ReservedFor indicates which entities are currently allowed to use the resource.
+   // Usually those are Pods, but other users are also possible.
+   //
+   // A scheduler must add a Pod that it is scheduling. This must be done in an
+   // atomic ResourceClaim update because there might be multiple schedulers working
+   // on different Pods that compete for access to the same ResourceClaim.
+   //
+   // kubelet will check this before allowing a Pod to run because a scheduler might
+   // have missed that step, for example because it doesn't support dynamic resource
+   // allocation or the feature was disabled.
+   ReservedFor []metav1.OwnerReference
+}
+
+type ResourceClaimPhase string
+
+const (
+    // The claim is waiting for allocation by the driver.
+    //
+    // For delayed allocation, the driver will wait for
+    // a selected node before it starts an allocation attempt.
+    ResourceClaimPending ResourceClaimPhase = “Pending”
+
+    // Set by the driver once the resource has been successfully
+    // allocated. The scheduler waits for all resources used by
+    // a Pod to be in this phase.
+    ResourceClaimAllocated ResourceClaimPhase = “Allocated”
+
+    // It can happen that a resource got allocated for a Pod and
+    // then the Pod cannot be scheduled onto the nodes where the allocated
+    // resource is available. The scheduler detects this and
+    // then sets the “reallocate” phase to tell the driver that it must
+    // free the resource. The driver does that and resets
+    // the ResourceClaimPhase back to "Pending".
+    ResourceClaimReallocate ResourceClaimPhase = “Reallocate”
+)
+
+type PodSpec {
+   ...
+   // ResourceClaims defines which ResourceClaims must be allocated
+   // and reserved before the Pod is allowed to start. The resources
+   // will be made available to those containers which reference them
+   // by name.
+   ResourceClaims []PodResourceClaim
```

Comment:
what happens if a claim is referenced that does not yet exist?

Reply:
The Pod cannot be scheduled.